### PR TITLE
makes hud images only show to mobs on the same z level to save on maptick

### LINKED ISSUE #63872: Taking off a medhud (item) doesn't remove the HUD (screen element)
Reporting client version: 514.1568

## Round ID:
[176088](https://scrubby.melonmesa.com/round/176088)

## Testmerges:
- [[MDB IGNORE] Adds progression traitors. Refactors uplink code in its entirety](https://www.github.com/tgstation/tgstation/pull/63588)
- [makes hud images only show to mobs on the same z level to save on maptick](https://www.github.com/tgstation/tgstation/pull/63558)

## Reproduction:

- Equip medical hud
- Change z levels
- Remove medical hud
- Still have the health bar above your head

image

### LINKED ISSUE #63901: Antag huds consistently don't work
## Round ID:

[176166](https://scrubby.melonmesa.com/round/176166)
Most recent Manuel rounds?

## Testmerges:

- [makes hud images only show to mobs on the same z level to save on maptick](https://www.github.com/tgstation/tgstation/pull/63558)

## Reproduction:

All antag huds consistently don't work.
Admin antag hud shows no antagonists - no traitors, revolutionaries, cultists, changelings, nothing
Team antag huds don't show any team-members for anyone involved, causing a large amount of confusion (especially on revolutionary rounds)
 
This has happened on both multi-z maps (icebox) and non-multi-z maps (deltastation)

### PR DESCRIPTION
**Pull request by Kylerace**
## About The Pull Request
makes hud images only show to players on the same z level and linked z levels above to optimize load on sendmaps a bit more. also cleans up the code a bit. eventually i want to make huds support currently inactive images to improve savings more but ill do that later
fixes #63872 
fixes #63901

## Why It's Good For The Game
so theres this step of the SendMaps profiler called "Check HUD/image visual contents". each tick, SendMaps iterates through every clients client.screen and client.images and checks each one for vis_contents and looking through a lot of sendmaps profiles it seems to take around 10% of all sendmaps varying a fair bit. for some reason images take about a fifth as long as screen movables to scan in this step buuuut a few days ago i realized that every ghost seems to have 
Screenshot_1672
images. and all of this it seems like is from atom_hud datums. it turns out that every single ghost gets: around 2 images of every mob in the world, 1 image of every door in the world, 1 image of every airlock in the world, 1 image of every bot path in the world, among presumably a few other painful examples. this is because they spawn with the advanced medical, security, and worst of all diagnostic huds. so every atom in the world those huds apply to are added to every ghost clients images list and increase maptick. each ghost also has around 60 screen movables from plane masters and such, but 60 * 5 is just 300, so screen objects are getting outweighed here by an order of magnitude. luckily most of the time ghosts are a minority of players, but on long rounds with lots of dead players i suspect the higher amounts of them can push maptick per person above 0.5

so by restricting it to images only in the players z level i (on local so the raw image count is lower than itd be on live) lowered it to around 300 on tramstation upper and lower levels and 500 on lavaland (before it was about 1600 on local)

### CODE REVIEW

**Comment by MrDoomBringer** [comment]
So I can't see someone's wanted status if they're down a set of stairs?

**Comment by MrMelbert** [comment]
If I'm reading this correctly: 
There's probably an argument to be made for doing this to the diagnostic hud (as it's the least useful hud), but not being able to see med-huds and sec-huds of people who are below you is pretty unfortunate for Multi-Z functionality. 

(Not being able to see the medhud of people who just got slammed by the tram and are on the ground dying below.)

**Comment by LemonInTheDark** [comment]
Surely we can group multiz levels together

**Comment by LemonInTheDark** [comment]
That or disable huds for ghosts by default or something

**Comment by Kylerace** [comment]
por que no los dos?

File: code/controllers/subsystem/spatial_gridmap.dm [hunk 1]
```diff
@@ -358,15 +360,15 @@ SUBSYSTEM_DEF(spatial_grid)
...
-	if(new_target.important_recursive_contents[RECURSIVE_CONTENTS_CLIENT_MOBS])
-		GRID_CELL_SET(intersecting_cell.client_contents, new_target.important_recursive_contents[SPATIAL_GRID_CONTENTS_TYPE_CLIENTS])
+	if(new_target.important_recursive_contents[SPATIAL_GRID_CONTENTS_TYPE_CLIENTS])
+		intersecting_cell.client_contents |= new_target.important_recursive_contents[SPATIAL_GRID_CONTENTS_TYPE_CLIENTS]
 
-		SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_ENTERED(RECURSIVE_CONTENTS_CLIENT_MOBS), new_target)
+		SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_ENTERED(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS), new_target.important_recursive_contents[SPATIAL_GRID_CONTENTS_TYPE_CLIENTS])
 
-	if(new_target.important_recursive_contents[RECURSIVE_CONTENTS_HEARING_SENSITIVE])
-		GRID_CELL_SET(intersecting_cell.hearing_contents, new_target.important_recursive_contents[RECURSIVE_CONTENTS_HEARING_SENSITIVE])
```
Comment (Kylerace) [inline code/controllers/subsystem/spatial_gridmap.dm:367]: i forgot that i left this in

Comment (Kylerace) [reply code/controllers/subsystem/spatial_gridmap.dm:367]: removed

File: code/modules/atmospherics/machinery/atmosmachinery.dm [hunk 2]
```diff
@@ -89,7 +89,7 @@
 /obj/machinery/atmospherics/Initialize(mapload)
 	if(mapload && name != initial(name))
 		override_naming = TRUE
-	return ..()
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/atmosmachinery.dm:92]: You summoned us for these

Comment (Kylerace) [reply code/modules/atmospherics/machinery/atmosmachinery.dm:92]: skill issue

Comment (LemonInTheDark) [reply code/modules/atmospherics/machinery/atmosmachinery.dm:92]: first time?

Comment (Ghilker) [reply code/modules/atmospherics/machinery/atmosmachinery.dm:92]: I'm gonna downvote the pr and there is nothing you can do about it!!!!!!?!

Comment (Kylerace) [reply code/modules/atmospherics/machinery/atmosmachinery.dm:92]: oh yeah so will i

Comment (LemonInTheDark) [reply code/modules/atmospherics/machinery/atmosmachinery.dm:92]: I hate you two

Comment (Kylerace) [reply code/modules/atmospherics/machinery/atmosmachinery.dm:92]: :trollface:

**Comment by RuneQuester** [comment]
SO this appears to break all HUDs, including cult and rev huds.

**Comment by Kylerace** [comment]
ill have to work on this later

**Pull request closed without merging**